### ISSUE TITLE
High resolution training

### ISSUE THREAD

**Issue by abbasmammadov** [ultralytics/yolov5#8843]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

First of all thank you very much for such an amazing work and your efforts to keep the model up to date with consistent updates.

Currently I am working on a dataset of images with the resolution of 2160*3360. There around 5 classes, and these objects are quite large in the all the images. 
First I used the below setting after processing my data into desired format:
!python3 train.py  --data splitted/data.yaml  --cfg yolov5s.yaml --weights yolov5s.pt 
Even on 60th epoch I terminated the code and I was able to achieve 88mAP, which was quite reasonable.

But to be honest I can not understand the image resolutions in this project/model. Above code due to the default setting was run on size of 640. But the results were accurate and at inference and validation time it is also quite good. 

Then I wanted to change --img-size to my data resolution but it takes only 1 argument and make it square of that size (e.g. 640*640). I changed it to the largest side i.e. 3360 but the training time was slow and size was 3360 3360 not 2160 3360. 
I am not sure I am doing it correctly since this part is quite confusing for me.

I would be glad if you can elaborate this size issues a bit more.
And also I would like to ask for advice that which settings should I train my data described above (imgs of 2160*3360 annotated according to the yolov5 format and object are large with at most 5 classes).

Thanks in advance for your attention!

**Comment by glenn-jocher** [comment]
@abbasmammadov 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by bit-scientist** [comment]
@abbasmammadov you would use larger models of yolo such as `l` or `x` and provide `--img-size 3360` with `--rect`

**Comment by abbasmammadov** [comment]
@bit-scientist  I solved the issue, but thank you very much for suggestion! 👍

**Comment by ChengGong0602** [comment]
@abbasmammadov @glenn-jocher 
How did u solve this issue?
I have similar issue with high resolution image dataset which is 5120*1400.
This is panorama images extracted from football matches video.
The goal of this training is to detect ball from the image.
The image resolution is high but ball is small.
When I have trained model with the standard size, it didn't detect ball as accurate.
What is the optimized image resize for training?

**Comment by glenn-jocher** [comment]
@ChengGong0602 Great to hear you've resolved your issue! 🎉

@bit-scientist For high-resolution images where the object of interest is small, like a football in a panoramic image, you might want to consider the following:

1. **Use a larger model**: YOLOv5 `l` or `x` models have more capacity to learn from high-resolution images.
2. **Adjust the `--img-size`**: Use a larger `--img-size` that is closer to your native image resolution, such as `--img-size 2048` or higher if your hardware allows it.
3. **Enable `--rect` training**: This flag allows for rectangular training, which can be beneficial for non-square images. It will adjust the smaller dimension of the image automatically while keeping the aspect ratio.
4. **Increase the number of anchors**: If the default anchors are not suitable for detecting small objects in your images, you might need to adjust them. You can use the `--autoanchor` flag to recalculate anchors based on your dataset.

Remember to monitor your validation mAP to ensure that the model is improving and not overfitting. If you're still not getting satisfactory results, consider using mosaic and mixup augmentations, or manually tuning hyperparameters to better suit small object detection.

Good luck with your training, and if you have further issues, don't hesitate to ask! 🚀
